Address "wide" alignment programatically

mtias opened the pull request:
This currently showcases the desired behaviour we want.

What I like is that it restricts side-effects that overflows and vw seem to have. What I dislike is that we need to dispatch when changes that affect content width happen (collapsing wp-admin sidebar manually).

jasmussen committed:
Remove full-width trick pending better one

This also adds a few improvements from the now deleted branch, https://github.com/WordPress/gutenberg/pull/799, notably:

- clearing for video blocks. The responsive trickery means we have to, or we will get odd behavior otherwise
- better collapsing margins fix for the toolbar.

mtias committed:
Calculate width and positioning for align:wide attr.

jasmussen committed:
Polish the toolbar centering math.

mtias committed:
Keep track of editor width values in redux state.

Update when sidebar is toggled.

aduth commented on editor/state.js:
```diff
@@ -303,6 +303,21 @@ export function isSidebarOpened( state = false, action ) {
 }
 
+export function editorWidth( state = null, action ) {
+	const layout = document.querySelector( '.editor-layout__content' );
```
These seem like arguments that should be passed in through action properties, not extracted by the reducer itself.

>Given the same arguments, it should calculate the next state and return it. No surprises. No side effects. No API calls. No mutations. Just a calculation.

http://redux.js.org/docs/basics/Reducers.html#handling-actions

aduth replied:
Also, every reducer is called for every action, so these query selectors are executed very often, and almost always needlessly so. At the very least we should check `action.type` before performing any queries.

mtias replied:
@aduth passing as action properties would mean we need to calculate wherever we dispatch the action.

aduth replied:
>which doesn't seem great.

In what way? We're not really computing much. We could persist references to the DOM nodes. Apparently `offsetWidth` [can force a reflow](https://gist.github.com/paulirish/5d52fb081b3570c81e3a#box-metrics), but if it's tied to resize event, shouldn't be a concern (or, rather, we're going to have to call it anyways).

mtias replied:
I mean more about cases where we may dispatch this from other places of the app, like the sidebar toggle. We shouldn't have to repeat the code there, I think.

aduth replied:
Hmm, might be middleware territory in an ideal solution then.

aduth reviewed (commented):
A related need for editor bounds (top, right, left) has been raised at https://github.com/WordPress/gutenberg/pull/839/files#diff-77cea533850e86cf5a0031b0acc4c0b2R40

jasmussen committed:
Fix issue with full width editor toolbar.

aduth commented on editor/layout/index.js:
```diff
@@ -11,26 +11,55 @@ import Header from '../header';
+class Layout extends wp.element.Component {
+
+	componentDidMount() {
+		window.addEventListener( 'resize', this.props.setEditorWidth );
```
We might consider debouncing this very noisy event.

BE-Webdesign replied:
> We might consider debouncing this very noisy event.

This ^

aduth commented on editor/layout/index.js:
```diff
@@ -11,26 +11,55 @@ import Header from '../header';
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'resize', this.props.setEditorWidth );
```
We should confirm that `mapDispatchToProps` is only called once for the lifetime of the component. Otherwise we run the risk of the reference here to `setEditorWidth` not being the same as the one assigned in `componentDidMount`, leaving zombie event listeners.

mtias committed:
Add wide and full-width options to images.

mtias committed:
Position toolbar in wide image option.

aduth commented on blocks/library/image/style.scss:
```diff
@@ -17,9 +17,14 @@
 	}
 
 	&[data-align="wide"] {
+		margin-left: -24%;
```
How well does this work on smaller viewports?

jasmussen replied:
Hmm yes this, for some reason, doesn't actually scale to the viewport. Replacing it with `5vw` is seems to behave as intended.

jasmussen committed:
Improve responsiveness.

There were some issues with centering blocks, due to the fact that it has a block mover on the left.

This commit simplifies that math, so the block mover doesn't actually affect the padding of the block, but rather uses negative space, making the math simpler.

Additionally, it adds some responsive fixes for the full wide toolbar positioning.

jasmussen commented:
@mtias I pushed a few responsive fixes to this branch, which will also benefit master. Do you think we can get this branch shipped to master this week? It'd be real nice to have.

mtias commented:
Yes, we'll get something this week. Discussing with @aduth the tradeoffs of the implementation.

aduth commented:
Somewhat related: #1023

mtias committed:
Completely different approach using CSS grid.

mtias committed:
Rework it all again with percentages.

mtias committed:
More elegant calculation of margins.

aduth commented on blocks/library/image/style.scss:
```diff
@@ -9,17 +9,28 @@
 		padding-left: 0;
 		padding-right: 0;
-		margin-right: -#{ $block-padding + $block-mover-margin };	/* Compensate for .editor-visual-editor centering padding */
+		box-sizing: border-box;
```
Shouldn't need this:

https://github.com/WordPress/gutenberg/blob/73bd4ab/editor/assets/stylesheets/main.scss#L17-L20

jasmussen replied:
👍

aduth commented on blocks/library/image/style.scss:
```diff
@@ -30,13 +41,39 @@
-			margin-left: auto;
-			margin-right: auto;
-		}
+			max-width: $visual-editor-max-width - ( ( $block-padding + $block-padding + $block-mover-margin ) * 2 );;
```
Double semi-colon ending.

aduth replied:
Should we move following comments before the `max-width` line?

jasmussen replied:
👍

I think the max-width was removed just now.

aduth commented on blocks/library/image/style.scss:
```diff
@@ -30,13 +41,39 @@
+			margin-right: $block-mover-padding-hidden + $block-padding;
+
+			@include break-small() {
+				width: 100%;
```
The position of the controls isn't quite right when at wide or full alignments. You can see it move left and right when toggling between an image "Align left" and "Align wide".

I think we need to assign `width: $visual-editor-max-width; max-width: none;`

jasmussen replied:
👍

aduth reviewed (commented):
Much happier to be solving this with CSS.

jasmussen committed:
Address responsive issues with emulated floats

We are emulating floats using margins. This works well enough, but has some responsive implications. This commit addresses those, by creating a mix-in that simplifies the breakpoints.

I couldn't quite figure out where to best place this mixin, and we also probably want to extract both the mixin, the variables, and even the left/right alignment code so that other blocks that utilize these don't have to copy the whole bit of code. But I wanted to push this so we can test, and then discuss.

jasmussen committed:
Address feedback.

Clean up double semicolons, better center toolbars, remove border-box.

jasmussen commented:
Addressed the feedback items, and also made the floats work with responsive. 

Take an extra look at https://github.com/WordPress/gutenberg/pull/812/commits/a4d712e4c7de1eb81b5bfab22f56741fac864c34 —  I wrote a small mix-in to simplify the responsive stuff a bit. Thankfully it's enough with three rules. 

Essentially we have to deal with responsive breakpoints that are unique to the editor, and unique to whether post settings is open or closed.

I could use advice on:

1. how the mixin can be improved
2. where it should live
3. per item 2, where it should live so that we can re-use the same code for any block that lets you left or right float an item.

I'm antsy to get this branch in though, so perhaps it'd be good if we could separate feedback in two groups. 1 —  improving the code so we can merge this, and 2 —  feedback we can address in a later PR. Thanks.

aduth commented:
>where it should live

We could move it to apply to [any `.editor-visual-editor__block`](https://github.com/WordPress/gutenberg/blob/378f441035533357f249d2417ab12ae1a5ca8ced/editor/modes/visual-editor/style.scss#L28). Is sorta how we're already treating it anyways with `getEditWrapperProps`.

jasmussen committed:
Move mixin, genericize alignments.

jasmussen commented:
Okay, took your advice and moved both the mixin and the alignments. 

@mtias what do you think? Ready to go?

jasmussen commented:
This branch, when merged, fixes #784 and #546.

mtias commented:
This is looking good. One side effect is we are losing the click-on-canvas-to-deselect.

jasmussen committed:
Hide movers and fix sticky menu

This commit hides the movers on floated images. We can look at reintroducing them for floats and full width images in a separate PR.

It also fixes another responsive issue where the sidebar is stickied.
